src/kernel/stage3.c

```diff
@@ -193,38 +218,51 @@ thunk create_init(kernel_heaps kh, tuple root, filesystem fs)
     return closure(heap_general(kh), startup, kh, root, fs);
 }
 
-closure_function(1, 1, status, kernel_read_complete,
-                 filesystem, fs,
+closure_function(2, 1, status, kernel_read_complete,
+                 filesystem, fs, boolean, destroy_fs,
                  buffer, b)
 {
-    add_elf_syms(b);
+    add_elf_syms(b, 0);
     deallocate_buffer(b);
-    destroy_filesystem(bound(fs));
+    if (bound(destroy_fs))
+        destroy_filesystem(bound(fs));
     closure_finish();
     return STATUS_OK;
 }
 
-closure_function(1, 2, void, bootfs_complete,
-                 kernel_heaps, kh,
+closure_function(4, 2, void, bootfs_complete,
+                 kernel_heaps, kh, tuple, root, boolean, klibs_in_bootfs, boolean, ingest_kernel_syms,
                  filesystem, fs, status, s)
 {
-    tuple root = filesystem_getroot(fs);
-    tuple c = children(root);
+    tuple boot_root = filesystem_getroot(fs);
+    tuple c = children(boot_root);
     assert(c);
-    table_foreach(c, k, v) {
-        if (!buffer_strcmp(symbol_string(k), "kernel")) {
-            kernel_heaps kh = bound(kh);
-            filesystem_read_entire(fs, v, heap_backed(kh),
-                                   closure(heap_general(kh),
-                                   kernel_read_complete, fs),
-                                   ignore_status);
-            break;
+    if (bound(klibs_in_bootfs)) {
+        init_klib(bound(kh), fs, boot_root);
```

Comment:
To complicate matters, if kernel symbols are not loaded, in general a kernel module cannot be loaded, because it wouldn't be able to resolve kernel symbols. So I think init_klib() should only be called after kernel symbols have been loaded. Which means that if the boot process doesn't go through stage2 (e.g. when booting via Frecracker), the kernel binary may need to be loaded even if not requested in the manifest file via `ingest_kernel_symbols`. This applies also if kernel modules are in the root filesystem...
Perhaps it would make sense to have a function in symtab.c that can be called to check whether kernel symbols have been loaded (it could be implemented by checking whether the symbol table covers an address known to belong to the kernel), so that the klib code could call this function and determine whether it's necessary to mount the boot filesystem and/or load the kernel binary.

Reply:
I think the klib_get_sym interface is fine, but I suggest that we don't bother with using the elf symtab. We could just as well have a list of functions to export to the klibs that are added to a table in init_klib. This would make for a more efficient lookup, which becomes a hash table lookup on an internalized symbol rather than a linear search through all kernel symbols with string compares. Operation would not be affected if either kernel symbols are not loaded (the scenario you describe above) or the kernel (or klibs) are stripped of symbols. And the -ffunction-sections issue won't be a problem, either, as exported functions will be referenced and thus linked into the kernel binary.

I'll update with an implementation and the other fixes you have provided.